[air/execution] Fix new execution backend for BOHB

krfricke opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

The new execution backend did not work with BOHB. The reason was a faulty check when we start actors. We should not eagerly start paused trials, as the scheduler may want to keep them paused for synchronous training. However, a leftover code piece allowed for trials that were once pending but then paused to be started. 

This PR fixes the bug by requiring trials to be strictly pending, and also makes sure that the `trial_to_run` is set to PENDING when chosen by the scheduler. BOHB tests are now enabled for the new execution backend.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

justinvyu reviewed (approved):
Nice! Added a few general questions.

Here's my summary of the problem and solution -- just wanted to clarify my understanding from the PR description:
1. Paused trials come from scheduler decisions. Once a scheduler decides a trial should be paused, it should basically be ignored by the Tune controller, until the scheduler decides that it's this trial's turn to run again.
2. Previously, it was possible to go directly from paused -> staged -> pending (set by `_schedule_trial_actor` -> running.
    - With this fix, it's now only possible to go from pending -> staged -> running.
    - So, the only route for a paused trial to resume would be: paused -> pending -> staged -> running.

Due to point 2, [should this line be an](https://github.com/ray-project/ray/blob/860a8c9c7b70daaea75e0172a4ccdd23f81146ad/python/ray/tune/execution/tune_controller.py#LL545C12-L545C12) `assert trial.status == PENDING` instead of setting it again? Sanity check that we only ever schedule an actor for a pending trial.

justinvyu commented:
Another (unrelated to PR) question while I'm at it.

What's the reason for removing the "original actor", then switching to the cached actor, where a trial already has a running actor? Maybe this is incurring a reset trial cost when we could just continue on the original actor?

https://github.com/ray-project/ray/blob/3139d4fcf1a98a421da109cf44770280e7d9e90e/python/ray/tune/execution/tune_controller.py#L517

krfricke commented on python/ray/tune/execution/tune_controller.py:
```diff
@@ -460,7 +461,7 @@ def _maybe_add_actors(candidates: List[Trial]):
                 # If the trial is part of the list, but not of the set,
                 # we just ignore it. Removing it from the list on status
                 # change is too expensive.
-                if trial not in (self._pending_trials | self._paused_trials):
+                if trial not in self._pending_trials:
```
We can theoretically have thousands of trials. The rationale here is that we add trials to `self._pending_trials`, which is a set and O(1), and to `self._pending_trials_list`, where insertion is O(1) and deletion is O(n). Instead of deleting from the list, we just iterate over the list and discard any objects that are not in the set.
The reason that we don't iterate over the set is because we want to keep the order the scheduler gave us intact.

justinvyu replied:
Oh, I mean this line, we're already incurring the O(n) deletion by popping from the front. Maybe we can change this to use a deque instead to have O(1) popleft.

https://github.com/ray-project/ray/blob/92f4fe725442a74705fe0c39d4cdf3f2d23d77c9/python/ray/tune/execution/tune_controller.py#L459

krfricke commented:
> Another (unrelated to PR) question while I'm at it.
> 
> For the following code, what's the reason for removing the "original actor", then switching to the cached actor? Maybe this is incurring a reset trial cost when we could just continue on the original actor?
> 
> https://github.com/ray-project/ray/blob/3139d4fcf1a98a421da109cf44770280e7d9e90e/python/ray/tune/execution/tune_controller.py#L517

The "original actor" refers to a tracked actor, and it usually has not been scheduled, yet (i.e. it's usually awaiting resources). If it were available, we would have started the trial earlier (on that actor).

krfricke commented on python/ray/tune/execution/tune_controller.py:
```diff
@@ -388,7 +388,7 @@ def _cleanup_trials(self):
         start = time.monotonic()
         while time.monotonic() - start < 5 and self._actor_manager.num_total_actors:
-            logger.debug("Waiting for actor manager to clean up final state")
+            # logger.debug("Waiting for actor manager to clean up final state")
```
Ah thanks!

krfricke commented:
> Nice! Added a few general questions.
> 
> Here's my summary of the problem and solution -- just wanted to clarify my understanding from the PR description:
> 
>     1. Paused trials come from scheduler decisions. Once a scheduler decides a trial should be paused, it should basically be ignored by the Tune controller, until the scheduler decides that it's this trial's turn to run again.
> 

Yes!

>     2. Previously, it was possible to go directly from paused -> staged -> pending (set by `_schedule_trial_actor`) -> running.

I think specifically the failure mode was: Trial was put into `self._pending_trials` and `self._pending_trials_list`. Then it was started, but via `trial_to_run`, not via the list, so it remained in `self._pending_trials_list`. It was then paused, so now it is in `self_pending_trials_list` and `self._paused_trials` - and we then tried to start it because it was still in `self._pending_trials_list`! Looks like this optimization bit us! I think it should be fine now 

>        
>        * With this fix, it's now only possible to go from pending -> staged -> running.
>        * So, the only route for a paused trial to resume would be: paused -> pending -> staged -> running.
> 
> 
> Due to point 2, [should this line be an](https://github.com/ray-project/ray/blob/860a8c9c7b70daaea75e0172a4ccdd23f81146ad/python/ray/tune/execution/tune_controller.py#LL545C12-L545C12) `assert trial.status == PENDING` instead of setting it again? Sanity check that we only ever schedule an actor for a pending trial.

Yes, that's a good point, and I've added the assertion. Thanks!
